pkg/ingester/ingester_v2_backfill.go

```diff
@@ -0,0 +1,499 @@
+	metrics     *ingesterMetrics
+
+	// dbs[0] is the one which would be overlapping with the main TSDB. "newer" TSDB.
+	// dbs[1] is the older TSDB among the two here. "older" TSDB.
+	// If the dbs[1] goes beyond the backfill age, it will be compacted
+	// and shipped and dbs[0] is moved to dbs[1]. During this process,
+	// there is only 1 TSDB ingesting old data, so there can be gaps till the compaction
+	// and shipping is going on.
+	// TODO(codesome): Avoid gaps by using a separate queue to compact and ship blocks.
+	dbs [2]*backfillTSDBWrapper
+
+	mtx sync.RWMutex
+	// We need this lock to safeguard force compactions from moving around of TSDBs between dbs array.
+	compactMtx sync.Mutex
+}
+
+type backfillTSDBWrapper struct {
+	db *userTSDB
+	// Time boundaries of the TSDBs in ms. Start is inclusive and End is exclusive.
+	start, end int64
+}
+
+func newBackfillTSDB(userID string, backfillAge time.Duration, metrics *ingesterMetrics) *backfillTSDB {
+	return &backfillTSDB{
+		userID:      userID,
+		backfillAge: backfillAge,
+		metrics:     metrics,
+	}
+}
+
+func (b *backfillTSDB) appender(
+	createNewTSDB func(userID, dbDir string, minBlockDuration, maxBlockDuration int64, reg *prometheus.Registry) (*userTSDB, error),
+	backfillBlocksDir func(userID string) string,
+) *backfillAppender {
+	return &backfillAppender{
+		createNewTSDB:     createNewTSDB,
+		backfillBlocksDir: backfillBlocksDir,
+		backfillTSDB:      b,
+	}
+}
+
+// backfillAppender is an appender to ingest old data.
+// This _does not_ implement storage.Appender interface.
+// The methods of this appender should not be called concurrently.
+type backfillAppender struct {
+	backfillTSDB                 *backfillTSDB
+	newerAppender, olderAppender *backfillRangeAppender
+
+	// Keeping this methods of Ingester here to avoid the appender having a reference to the ingester.
+	createNewTSDB func(userID, dbDir string, minBlockDuration, maxBlockDuration int64, reg *prometheus.Registry) (*userTSDB, error)
+	// We are keeping the method instead of string for backfill dir because if there is a surge in backfill writes,
+	// it will create a new string for every push while we want it only to open a new TSDB.
+	backfillBlocksDir func(userID string) string
+}
+
+type backfillRangeAppender struct {
+	app        storage.Appender
+	db         *userTSDB
+	start, end int64
+}
+
+func newBackfillRangeAppender(db *backfillTSDBWrapper) *backfillRangeAppender {
+	return &backfillRangeAppender{
+		app:   db.db.Appender(),
+		db:    db.db,
+		start: db.start,
+		end:   db.end,
+	}
+}
+
+// add requires the samples to be within the backfill age.
+// Note: This method should not be called concurrently.
+func (a *backfillAppender) add(la []client.LabelAdapter, s client.Sample) (err error) {
+	app, db, err := a.getAppender(s)
+	if err != nil {
+		return err
+	}
+	startAppend := time.Now()
+
+	cachedRef, cachedRefExists := db.refCache.Ref(startAppend, client.FromLabelAdaptersToLabels(la))
```

Comment:
To my understanding, `db.refCache` is shared across all TSDBs (main + 2 backfill) but is it safe to be shared?

Reply:
The backfill TSDBs themselves are `userTSDB` and have their own `refCache` (each backfill TSDB getting one). They don't have any access to the main TSDB.